### ISSUE TITLE
Ability to create invalid records

### ISSUE THREAD

**Issue by vorner** [bluejekyll/trust-dns#320]
Hello

I noticed it is possible to create invalid record. Eg, this compiles (and runs), but produces invalid data when serialized (claims to be A, but has too much bytes).

```rust
let mut record = Record::with(name, RecordType::A, 3600);
record.set_rdata(RData::AAAA("db8::1".parse().unwrap());
```

Does it make sense to be able to create a record with different rdata than its type? Does it make sense to create an „empty“ record? Wouldn't it be better to be able to create only valid records (directly fill them with rdata, derive the record type from the rdata, etc).

**Comment by bluejekyll** [comment]
> Does it make sense to be able to create a record with different rdata than its type?

There was a point where I was reusing common RData (like Name) across different RecordTypes, where this would have been valid. I've since moved away from that and started explicitly using separate RData types, with single values as the actual data section of the RData type. So there should be (and I think there is now) a 1-1 relationship between RecordType and RData. We could hide RecordType entirely I think at this point. RData does have a method to get the RecordType from it.

> Does it make sense to create an „empty“ record?

Depends on what you mean by empty. There are cases where the `Null` RData type is valid, but they are mostly edge conditions.

> Wouldn't it be better to be able to create only valid records?

Yes. We should definitely add some validation on the RData to restrict it to RecordType. At one point I toyed with trying to make a generic type parameter to catch this at compile time, but I don't think it's easily done without adding more intermediate types. So yes, I think we could drop RecordType from the Record constructor all together, possibly each from the Record itself, and fix locations as necessary.

**Comment by bluejekyll** [comment]
> What I was thinking would be to have to pass all the information at construction time. That way it wouldn't be easy to create „broken“ ones. But maybe I'm missing some edge case.

Yes, I think this would be a good solution. If I were writing this today, I might opt to do that. When I wrote it originally, I don't think all the potential use cases were obvious to me at the time, so I opted for more flexibility. I'm all for restricting this more now, and requiring new Records to be created if you want to change shape...

**Comment by bluejekyll** [comment]
This was fixed in #674
